### PR TITLE
better default capacity for str::replace

### PR DESCRIPTION
**Pull request by LaihoE**
Adds smarter capacity for str::replace in cases where we know that the output will be at least as long as the original string.

### CODE REVIEW

**Comment by LaihoE** [comment]
Could someone trigger a rust-timer run?

File: library/alloc/src/str.rs [hunk 1]
```diff
@@ -280,8 +279,16 @@ impl str {
                 return unsafe { replace_ascii(self.as_bytes(), from_byte, *to_byte) };
             }
         }
-
-        let mut result = String::new();
+        let to_len = to.as_bytes().len();
+        // Set result capacity to self.len() when from.len() <= to.len()
+        let default_capacity = match from.as_utf8_pattern() {
+            Some(Utf8Pattern::StringPattern(s)) if s.len() <= to_len => self.len(),
+            Some(Utf8Pattern::CharPattern(c)) if c.encode_utf8(&mut [0; 4]).len() <= to_len => {
```
Comment (joboet) [inline library/alloc/src/str.rs:286]: There's a dedicated method for this:
```suggestion
            Some(Utf8Pattern::CharPattern(c)) if c.len_utf8() <= to_len => {
```

**Review by joboet — APPROVED** [review]
Thank you, that's a good idea.
r=me after addressing the nit

**Comment by klensy** [comment]
Can comment https://github.com/rust-lang/rust/pull/130223#issuecomment-2395007697 be updated here?

**Comment by LaihoE** [comment]
> Can comment [#130223 (comment)](https://github.com/rust-lang/rust/pull/130223#issuecomment-2395007697) be updated here?

Done

File: library/alloc/src/str.rs [hunk 2]
```diff
@@ -280,8 +279,14 @@ impl str {
                 return unsafe { replace_ascii(self.as_bytes(), from_byte, *to_byte) };
             }
         }
-
-        let mut result = String::new();
+        let to_len = to.as_bytes().len();
```
Comment (joboet) [inline library/alloc/src/str.rs:282]:
```suggestion
        let to_len = to.len();
```

**Review by joboet — APPROVED** [review]
r=me with the nit fixed.

**Pull request merged**